### DISCUSSION TITLE
Post quantum crypto

### DISCUSSION THREAD

**Discussion by Paragis** [FiloSottile/age#231] [Spec feedback]
Given that age aims to be the encryption tool for the future, it should include (asymmetric) crypto that will survive the expected arrival of quantum computers.

openssh already has this implemented:
> ... based on a combination of Streamlined NTRU Prime 4591^761 and X25519.
https://www.openssh.com/releasenotes.html

Alternatively the NIST Competition for post-quantum crypto is getting to the final round in summer 2020
https://en.wikipedia.org/wiki/Post-Quantum_Cryptography_Standardization

**Comment by Macil** [comment]
Do we know if the public keys for these systems are still short enough to use as command-line arguments comfortably, or would a post-quantum age need arguments that are files containing the public keys? I know some post-quantum systems have very large public keys.

**Reply by qnfm** [reply]
The length of encoded kyber512's public key is 1286 char. To trim its secret key size, I only store its key seed which is 32 bytes

**Reply by keisentraut** [reply]
The key sizes of sntrup761 are as following (see [here](https://ntruprime.cr.yp.to/security.html)):
- secret key: 1763 bytes
- public key: 1039 bytes

With Bech32 encoding, this will expand by +60% and result in very large recipients/identities.

**Reply by qnfm** [reply]
Sorry about that,the key seed size "32 bytes" was ture for kyber512's cpa-public key encryption,which I used to integrate into my first two version(<rc.3).
The key sizes before being encoded by bech32 in rc.3,which calls to cca-kyber512-kem,  are as following:
- key seed: 64bytes
- public key:800bytes

**Reply by qnfm** [reply]
> The key sizes of sntrup761 are as following (see [here](https://ntruprime.cr.yp.to/security.html)):
> 
>     * secret key: 1763 bytes
> 
>     * public key: 1039 bytes
> 
> 
> With Bech32 encoding, this will expand by +60% and result in very large recipients/identities.

sntrup761's public key size is 1158 from [this](ntruprime.cr.yp.to)

**Reply by ghost** [reply]
sntrup has known yet to be verified attack venues that exploit the underlying mathematical properties of lattices not ruled out by proofs, potential systemic risks.

#  To repeat - the underlying mathematical properties of lattices

Thanks.

**Comment by cyb3rz3us** [comment]
Given that the exchange method suggested above is still experimental, IMO this seems a bit premature for insertion into 'age' at present.

**Reply by ghost** [reply]
Welcome to the world of crypto in Go. Stay tuned for entertainment.

Thanks.

**Comment by Timo67235** [comment]
The 3. round of the NIST Post-Quantum Cryptography Standardization has already revealed a number of potential candidates for future post-quantum algorithms https://en.wikipedia.org/wiki/Post-Quantum_Cryptography_Standardization

The remaining algorithm all seem to have significantly larger key sizes, wich means that no matter which algorithms will be chosen in the end, large keys sizes will remain an issue would have to be dealt with.

With ed25519 breaking as soon as quantum computers with 3000 qubits arrive, modern crypto tools like age will need support for post quantum crypto rather sooner then later.

**Comment by keisentraut** [comment]
Just a small remark: Password-based encryption is already post-quantum secure because it only uses scrypt for key derivation and ChaCha20-Poly1305 for (authenticated) encryption. You can expect 128bit of post-quantum security against a powerful quantum computer who can run Grover's algorithm on the complete 256bit search space.

**Reply by markdascher** [reply]
I'm curious how confident we all are in this statement? The part that makes me wonder is the [128-bit file key](https://github.com/C2SP/C2SP/blob/main/age.md#file-key).

I'm unfamiliar with the actual math myself–just that 128-bit is bad and 256-bit is good. 😄

Found a similar question asked [here](https://crypto.stackexchange.com/questions/76332/is-using-aes-256-with-a-key-of-less-than-256-bits-of-entropy-still-quantum-resis) but with only a handful of comments.

**Reply by keisentraut** [reply]
You are right and my statement was wrong. You can expect only 64bit of post-quantum security against a powerful quantum computer. AFAIK, this should still be a much harder task than breaking the DLP problem for curve25519 with a quantum computer, i.e. breaking the default age recipient.

**Reply by FiloSottile** [reply]
I looked into this and according to more realistic models, the 128-bit key might be ok. https://twitter.com/FiloSottile/status/1544680619346694147

**Comment by agucova** [comment]
OpenSSH moved to make it a default on 9.0:
> [ssh(1)](https://man.openbsd.org/ssh.1), [sshd(8)](https://man.openbsd.org/sshd.8): use the hybrid Streamlined NTRU Prime + x25519 key
   exchange method by default ("sntrup761x25519-sha512@openssh.com").
   The NTRU algorithm is believed to resist attacks enabled by future
   quantum computers and is paired with the X25519 ECDH key exchange
   (the previous default) as a backstop against any weaknesses in
   NTRU Prime that may be discovered in the future. The combination
   ensures that the hybrid exchange offers at least as good security
   as the status quo.

**Reply by qnfm** [reply]
Here is my experimental implementation(for demo usage):
https://github.com/qnfm/age

**Reply by qnfm** [reply]
It could be compatible with original version.I just tweakd it for myself

**Reply by keisentraut** [reply]
@qnfm This is funny, I did something similiar. 
I wrote a age plugin which does a hybrid encryption with x25519 and sntrup761. However, it comes with two downsides: You cannot convert a secret key to a public key and the recipients/identities are very long. But it should be as least as safe as x25519 alone. My plugin only runs with rage so far, because age does currently not support any plugins. See https://github.com/keisentraut/age-plugin-sntrup761x25519 .

**Reply by qnfm** [reply]
For the first issus, due to cloudflare's excellent work, there is an api to generate key pair deterministically from the given seed:
`(pk,sk)<-Gene(seed)`
And yeah, it is inevitable according to all round3 kem finalists, there will be at least one key containing a lot of info(e.g matrix). So I'm thinking about whether qrcode can be another way to represent public keys

**Comment by FiloSottile** [comment]
Now that NIST selected a post-quantum KEM for standardization, I will start to experiment with a Kʏʙᴇʀ512+X25519 plugin.

There are definitely UX issues around the public key size that will need solving, and I don't expect it to exit an experimental state until the NIST standard is published in 2024. Note that there might be patent issues with using Kʏʙᴇʀ before a NIST standard is available.

The good news is that we might not need to version the protocol to extend the file key, though. https://twitter.com/FiloSottile/status/1544680619346694147

**Reply by artemislena** [reply]
T.: W long key sizes, people're *gonna* use hashes for easier identification n verification (e.g. as for the latter, people might download the key from an official source, then compare its hash w, say, one that's specified on one or several third-party sources like https://artemislena.eu/services/verify.html). We'd suggest'ving some hashing stuff built in (or at the very least, a standard scheme being recommended), prolly based on BLAKE2b or SHA3 w sufficient digest size (256 bits or more?). 'f there's no official standard on hashing, people might mix these as well as less secure hashing functions, which'd worsen UX (n security).

**Reply by artemislena** [reply]
T.: https://blog.cr.yp.to/20231003-countcorrectly.html Ya may wanna reconsider Kyber512 specifically? Larger Kyber keys (or just using NTRU instead) might be fine, but accordinga this, Kyber512 doesn't even reach AES128's security margin.

**Reply by JosiahBSharkey** [reply]
Kyber was chosen because it was one of the least secure options like aes being chosen as standard instead of threefish or serpent that are more secure in every way but slower than hardware accelerated aes. For offline encryption you should not use the least secure standardized encryption because they are made to make people not notice the time it takes for online encryption. aes should also be replaced with serpent 256 using a tweakable stream cypher mode and kmac or threefish 1024 authenticated tweakable stream cypher because they are both better in every way than aes and just a little bit slower than hardware accelerated aes especially for offline encryption. Post quantum encryption is new and less tested than pre quantum and decryption time doesn't matter because it is offline so the best option would be a cascade of multiple hybrid post quantum algorithms. Scrypt should also be replaced with argon2 then the key size doesn't matter you can just put it in a file with more than 30 seconds of argon2 and a long password then you can store your secret keys with the secrets. This is what I do already but I keep my secrets in a luks loopback file and my keys in a luks loopback file inside of that because serpent is more secure than aes and it gives me a cascade of post quantum algorithms using a post quantum plugin to access secrets and a cascade of serpent and aes to protect secrets and I can store everything publicly without losing security because of a cascade of serpent protected with more than 30 seconds of argon2 with long passwords or sss locked with clevis and tang over wireguard psk for post quantum security, because sss is post quantum as long as you don't have threshold keyshares, to access secret keys and I sign the luks loopback files with sphincs+ so they can't be trusted if it is modified then I can use tools that require age without getting hacked by the NSA because they think I am a spy or a criminal because my opsec is so good they don't know what I am doing without hacking me even though they can trace me because there aren't any post quantum anonymity nets but they can't break the post quantum privacy of ssh session keys but they can steal the authentication key or host key and use those to hack you because they are pre quantum unless you run it over a post quantum VPN like wireguard psk

**Reply by ghost** [reply]
> Oct 9, 2023 comment extract - aes should also be replaced with serpent 256 using a tweakable stream cypher mode and kmac or threefish 1024 authenticated tweakable stream cypher because they are both better in every way than aes and just a little bit slower than hardware accelerated aes especially for offline encryption.

use the last version of Threefish 1.3

<details><summary>Privacy or ...</summary>
<p>


the last part about anonymity is a pointless war of attrition against the manifestly inevitable comedic fall. if you want privacy do something extreme, hide everything, isolate isolate isolate and never live in peace for half a `moment of a moment` (moment is an instant or is it?) OR ...

</p>
</details> 

Thanks

**Comment by JosiahBSharkey** [comment]
any encryption software that isn't post quantum is broken the only reason I can use age is because it has post quantum age plugins I have had a persistent active attacker that has broken into my house and covertly stolen things like encrypted drives and I was running ssh ed25519 and passwords turned off and tor authorized clients ed25519 and after a month I would have malware on my computer and that was before I have evidence of them stealing from me or breaking into my house this has been happening for a few years now so that means that a nation state that thinks I'm trying to hide something from them with my anonymity but I just want privacy for everything I don't tell people has been attacking me with a pretty slow quantum computer because it takes it a month to break 2 ed25519 keys

**Reply by ghost** [reply]
please read https://github.com/FiloSottile/age/discussions/231#discussioncomment-12346171

**Comment by lirc572** [comment]
Now that [the NIST PQC standards are published](https://www.nist.gov/news-events/news/2024/08/nist-releases-first-3-finalized-post-quantum-encryption-standards), shall we start working on [the new PQC recipient type](https://github.com/FiloSottile/age/issues/578#issuecomment-2171252663)?

Other references:

- [ML-KEM X25519 support in OpenSSH 9.9](https://www.openssh.com/releasenotes.html#9.9)
  - [IETF draft: PQ/T Hybrid Key Exchange in SSH](https://datatracker.ietf.org/doc/html/draft-kampanakis-curdle-ssh-pq-ke-04)

**Reply by qnfm** [reply]
Agree

**Comment by dksslq** [comment]
Now, golang 1.24 starting supported ML-KEM in its standard library: [New crypto/mlkem package](https://go.dev/doc/go1.24#crypto-mlkem)

**Comment by david415** [comment]
What is currently stopping you from adding post quantum cryptography to "age"?
Obviously you can do it and nothing is actually stopping you... so all I'm really asking here is for your rationale for not providing the people with a more useful tool. My country is currently falling to fascism and we need tools like this in order to properly organize resistance. PQ Crypto would be a minimum baseline for considered use in a public cryptography setting.
What's the rationale?

**Reply by FiloSottile** [reply]
I invite you to take some time to reflect on how awful this comment is, and what led you to posting it.

Against my best judgement, I will add that what kept me from implementing what you demand in this free open source tool (which supports third-party plugins) includes implementing ML-KEM in the Go standard library and helping specify hybrid KEMs in HPKE.

(Also, waiting for the most inane CFRG bikeshed session on what strings to use to domain separate hybrid KEM instances.)

**Reply by david415** [reply]
Upon further reflection I do NOT see my own comment above as being demanding. But I guess the accusation is your way of saying that you don't like my question. Why the change of heart with regards to PQ? Perhaps it was all of this in your recent blog post? https://words.filippo.io/crqc-timeline/ Or did you just now figure out what strings to use to name things?

**Reply by ThinkChaos** [reply]
Yes you're demanding and condescending. Filo has obviously spent a lot of time and effort both in standardization and implementation for PQ crypto. As you've read the blogpost I'm sure you know it answers your questions, and that age 1.3.0 has PQ support. So rejoice that you can now use PQ age, and stop being toxic to someone that has actually done what you wanted all along. If you're not happy with timelines, just learn crypto and help.

**Reply by Malix-Labs** [reply]
https://github.com/FiloSottile/age/discussions/231#discussioncomment-16467330 from @david415 :
> Or did you just now figure out what strings to use to name things?

This is an example of a low-value non-contributor under-achieving high-standard user being a dick.
Have you said thank you once?

3 days ago, https://github.com/nvim-treesitter/nvim-treesitter (13.6k stars) got archived because of a very similar comment and behavior : https://github.com/nvim-treesitter/nvim-treesitter/discussions/8627#discussioncomment-16440743 (also see [the related X post](https://x.com/Malix_Labs/status/2040360074346947069?s=20))

I do not want to take any chance of such event happening to one of the project I use and love;
So @david415, the age project and Filo doesn't owe you anything, and you're absolutely free to use another encryption product (lucky you, there are alternatives!).

There is also no reason not to block you from interacting with this project anymore, which is what I would enact if I was a maintainer.
You also can be glad he is so patient and putting up with people like you.

Thanks again to him for implementing PQC for free to a great tool I use basically every day <3

**Reply by FiloSottile** [reply]
I have blocked @david415, so we should now let this thread rest, since he can't respond anymore.

**Comment by bordam** [comment]
If anyone lands here, post-quantum encryption has been implemented in [version 1.3.0](https://github.com/FiloSottile/age/releases/tag/v1.3.0)! Thanks a lot, @FiloSottile

**Reply by reneleonhardt** [reply]
Awesome! ❤️ It will be interesting to see when examples for long term security will be added (ML-KEM-1024 + X25519 = "X-Bow", or MLKEM1024-P384 with FIPS 202 allowing SHA3-384 and SHA3-512).

https://datatracker.ietf.org/doc/draft-ietf-hpke-pq/

> ### 4.  Hybrid KEMs with ECDH and ML-KEM
> MLKEM1024-P384:  ML-KEM-1024 and P-384

**Reply by FiloSottile** [reply]
ML-KEM-768 is perfectly fine for long-term security.

ML-KEM-1024 hybrids are mostly being specified for compliance reasons, because in the 90ies the US government became enamored with the concept of "security levels" and never recovered, even if these days all practical systems target the same security level: "secure unless completely new (and hence unpredictable) math is discovered"

https://www.imperialviolet.org/2014/05/25/strengthmatching.html and https://eprint.iacr.org/2019/1492 are good reads.

**Reply by Malix-Labs** [reply]
Banger, thanks Filo

**Reply by handpickencounter** [reply]
Erosion of the security margin is something which can occur with cryptographic primitives which rely on mathematics. RSA was eroded in this fashion. but then if age was concerned about providing long lasting security it would of started by using curve448 instead of 25519.
It's unfortunate that if you want secure longevity you should be using large SSH RSA keys (with an additional age layer for the PQC, until such a time as another tool emerges which uses MLKEM-1024 and McEliece together).

**Reply by FiloSottile** [reply]
ML-KEM-768 is already a conservative choice over -512, which would be nominally secure.

Anything above 128 bits of security is anyway speculation about math that does not exist or numerology.

I doubt you will find a cryptographer that has more confidence in any RSA size than in X-Wing, but if you have personal preferences, age is indeed not the right tool.

**Reply by learnwayback** [reply]
I don't understand why ChaCha20Poly1305 is used if 128-bit security is the aim? Any efficiency gained by opting not to target 256-bit security in the public key crypto is wiped away by the opportunity cost of not using AES128 which is going to be accelerated on the vast majority of hardware (and a 1 bit switch could have a fallback to chacha8 or something)
